Add assign argument to torch.Tensor.module_load

mikaylagawarecki opened the pull request:
Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121158
* #121157

mikaylagawarecki committed:
Add assign argument to torch.Tensor.module_load

[ghstack-poisoned]

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -2103,7 +2103,8 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
             If :attr:`assign` is ``True`` the optimizer must be created after
-            the call to :attr:`load_state_dict`.
+            the call to :attr:`load_state_dict` unless
+            :func:`~torch.__future__.get_swap_module_params_on_conversion` is ``True``.
```
Would appreciate feedback on how to make the behavior of `assign` when swap is set more clear in the docstring for the argument :)

albanD replied:
The way I think about this is: one preserves the properties of the Tensor currently in the model while the other preserves the properties of the Tensor currently in the state dict.

mikaylagawarecki replied:
Agreed! And we do document it as such!

 > assign (bool, optional): whether to assign items in the state
                dictionary to their corresponding keys in the module instead
                of copying them inplace into the module's current parameters and buffers.
                When ``False``, the properties of the tensors in the current
                module are preserved while when ``True``, the properties of the
                Tensors in the state dict are preserved.
                Default: ``False`

I do feel that if I was not familiar with the internals, it would be a little strange to reason about what swapping means in conjunction with `assign` though. In particular, do you think there is a better way to word

> assign items in the state dictionary to their corresponding keys in the module

for the swap case, or do you think this seems fine as it is

Also one small edge case is: do we consider `requires_grad` a property of the tensor? :P

albanD replied:
We should definitely callout the requires_grad special behavior here.

I do think that the doc can be simplified here. We go a bit too much in the details of how its implemented maybe? If we do need a 100% accurate description, we most likely need a longer form comment anyways that mentions how each is implemented.
So maybe I would just remove the first sentence there and only keep "When False, the properties of the tensors in the current module are preserved while when True, the properties of the Tensors in the state dict are preserved.".

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -2046,16 +2046,9 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                        if use_swap_tensors:
                             param_requires_grad = param.requires_grad
-                            new_input_param = param.module_load(input_param)
+                            new_input_param = param.module_load(input_param, assign=assign_to_params_buffers)
```
should we add an assert to make sure that `__torch_function__` overrides conform to the definition for `assign` or should we let the torch function handler do anything?

albanD replied:
How would you check that it conforms to it?

mikaylagawarecki replied:
Something like 

```
if assign:
         assert type(new_input_param) == type(input_param)
else:
         assert type(new_input_param) == type(param)
 ```
 
 But perhaps this is too restrictive

albanD replied:
This feels too restrictive indeed, We did mention that in the subclass case, even when assign=False, we do say that you can change the type there.

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

albanD commented on torch/_tensor.py:
```diff
@@ -727,12 +727,17 @@ def module_load(self, other):
+                Tensor.module_load, (self, other), self, other, assign=assign
+            )
+
+        if not assign:
+            return self.copy_(other).detach()
+        return other.detach()
```
nit: if/else might be slightly easier to read for such as simple case

albanD commented on torch/_tensor.py:
```diff
@@ -727,12 +727,17 @@ def module_load(self, other):
+            )
+
+        if not assign:
+            return self.copy_(other).detach()
```
We need to update the doc about the defaul implementation above

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

albanD commented on torch/nn/modules/module.py:
```diff
@@ -2103,20 +2103,22 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
-                Tensors in the state dict are preserved.
+            assign (bool, optional): When ``False``, the properties of the tensors
+                in the current module are preserved while when ``True``, the
+                properties of the Tensors in the state dict are preserved. Additionally,
+                the ``requires_grad`` of :class:`~torch.nn.Parameter`s in the
+                ``state_dict`` are preserved, but tensors in the ``state_dict``
+                will adopt the ``requires_grad`` of the corresponding parameter
+                in the module.
```
```suggestion
                properties of the Tensors in the state dict are preserved. The only exception is
                the ``requires_grad`` field of :class:`~torch.nn.Parameter`s for which the 
                value from the module is preserved.
```

Maybe simpler phrasing?

albanD reviewed (approved):
Sounds good!
Just a proposal for simpler phrasing

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add assign argument to torch.Tensor.module_load"

Make `torch.__future__.get_swap_module_params_on_conversion() == True` account for `assign` argument to `nn.Module.load_state_dict`

Similar to when `torch.__future__.set_swap_module_params_on_conversion()` is `False`, `assign=True` means that we do not incur a `self.copy_(other)` and the properties of `other` will be preserved




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
